File: hw2/hw2.R
source("rwishert.R")
source("../R_Functions/plotPost.R",chdir=TRUE)

# Sufficient Stats:
Y <- as.matrix(iris[,1:2])
Y.bar <- apply(Y,2,mean)
c.rows <- t(apply(Y,1,function(y) y - Y.bar))
C.mat <- matrix(0,ncol(Y),ncol(Y))
for (i in 1:nrow(c.rows)) {
  C.mat <- C.mat + c.rows[i,] %*% t(c.rows[i,])
}

# Prior Params:
s <- 1
r <- 10
k <- length(Y.bar)
n <- nrow(Y)
S <- diag(k)
m <- Y.bar

# Posterior Params:
m.post <- (s*m + n*Y.bar) / (s+n) 
s.post <- s+n
r.post <- r+n
S.post <- S + C.mat + s*n/(s+n) * (Y.bar-m) %*% t(Y.bar-m)
iS.post <- solve(S.post) # IMPORTANT!!!

B <- 100000

postpred.y <- lapply(as.list(1:B),function(i) rniw.postpred(m.post,s.post,r.post,iS.post,brief=FALSE))

post.mu <- t(sapply(postpred.y,function(xx) xx$mu))
post.mu.info <- t(apply(post.mu,2,quantile,c(.025,.5,.975)))

post.S <- lapply(postpred.y, function(xx) xx$S)
post.S.info <- rbind(
quantile(sapply(post.S,function(i) i[1,1]),c(.025,.5,.975)),
quantile(sapply(post.S,function(i) i[2,2]),c(.025,.5,.975)),
quantile(sapply(post.S,function(i) i[1,2]),c(.025,.5,.975))
)

out.summary <- rbind(post.mu.info[1:2,],post.S.info)
out.summary
rownames(out.summary) <- c("Sepal Length", "Sepal Width", "SL Var", "SW Var", "SLW CoVar")
out.summary

pp <- t(sapply(postpred.y,function(xx) xx$postpred))
plot(Y[,1],Y[,2],xlab="Sepal Length",ylab="Sepal Width",pch=20,col="grey",cex=3)
post.mean <- apply(pp[,1:2],2,mean)
points(post.mean[1],post.mean[2],cex=5,col="cornflowerblue",pch=20)
plot.contour(pp[,1:2],add=TRUE,col="cornflowerblue")

File: hw2/jl/hw2.jl
println("Loading Libraries...")
using RCall, Distributions, PyPlot, KernelDensity
println("Done loading libraries...")

R"
source('../../R_Functions/plotPost.R',chdir=TRUE)

Y <- as.matrix(iris[,1:2])
Y_bar <- apply(Y,2,mean)
c_rows <- t(apply(Y,1,function(y) y - Y_bar))
C_mat <- matrix(0,ncol(Y),ncol(Y))
for (i in 1:nrow(c_rows)) {
  C_mat <- C_mat + c_rows[i,] %*% t(c_rows[i,])
}

# Prior Params:
s <- 1
r <- 10
k <- length(Y_bar)
n <- nrow(Y)
S <- diag(k)
m <- Y_bar

# Posterior Params:
m_post <- (s*m + n*Y_bar) / (s+n) 
s_post <- s+n
r_post <- r+n
iS_post <- solve(S) + C_mat + s*n/(s+n) * (Y_bar-m) %*% t(Y_bar-m)
S_post <- solve(iS_post) # IMPORTANT!!!

B <- 100000
"
@rget Y Y_bar C_mat m_post s_post r_post iS_post S_post B
var_names = ["Sepal Length", "Sepal Width"]
plot_posts = R"plot.posts"
rplot = R"plot"
rapply = R"apply"
sapply = R"sapply"

function rniw(m,s,r,iS,brief=true)
  iW = rand(InverseWishart(r,iS))
  mu = rand(MultivariateNormal(m,iW/s))
  postpred = rand(MultivariateNormal(mu,iW))
  out =  brief ? postpred : (mu,iW,postpred)
  out
end

# using PyPlot, KernelDensity, Distributions
function biv_contour(x,y,c="grey",add=false)
  if add KernelDensity.PyPlot_init() end
  k = kde((x,y))
  contour(k,colors=c)
end

par = m_post, s_post, r_post, iS_post, false
@time par_list = [par for i in 1:B]
println("Sampling...")
@time out = map(x -> rniw(x[1],x[2],x[3],x[4],x[5]), par_list);

post_mu = map(x -> float(x), 
              hcat(map(x -> x[1][1], out), 
                   map(x -> x[1][2], out)))

@time post_mu_info = rcopy(rapply(post_mu, 2, x -> quantile(x,[.025,.5,.975])))'
@time post_S = map(x -> x[2], out)
@time post_S_info = hcat(
  quantile(rcopy(sapply(post_S, S -> S[1,1])),[.025,.5,.975]),
  quantile(rcopy(sapply(post_S, S -> S[2,2])),[.025,.5,.975]),
  quantile(rcopy(sapply(post_S, S -> S[1,2])),[.025,.5,.975])
)'

out_summary = vcat(post_mu_info, post_S_info)

pp = map(x -> float(x), 
         hcat(map(x -> x[3][1], out), 
              map(x -> x[3][2], out)))

scatter(mean(pp[:,1]),mean(pp[:,2]),s=500,c="cornflowerblue",edgecolors="none")
scatter(Y[:,1],Y[:,2],s=100,c="grey",edgecolors="none")
biv_contour(pp[:,1],pp[:,2],"grey",true)
#=
include("hw2.jl")
plot_posts(Y,names=var_names) 
plot_posts(post_mu[Int(B*.95):end,:],names=var_names)
=#
